v2/devdoc/logging.md

```diff
@@ -0,0 +1,221 @@
+# `c-logging` requirements
+
+## Basic functionality
+
+`c-logging` shall expose APIs allowing the user to log events that can be used for observing the software activity, debugging, etc.
+
+The exposed API shall allow creating contexts both dynamically and on the stack and using those contexts for logging.
+
+The API shall allow easy formatting of values to be placed in the contexts by using printf-like formatting.
+
+```
+void logger_log(LOG_LEVEL log_level, LOG_CONTEXT_HANDLE log_context, const char* file, const char* func, int line_no, const char* format, ...);
+
+#define LOGGER_LOG(log_level, log_context, format, ...) \
+    logger_log(log_level, log_context, __FILE__, __FUNCTION__, __LINE__, format MU_IFCOMMALOGIC(MU_COUNT_ARG(__VA_ARGS__)) __VA_ARGS__)
+
+// macro that can be used to create a dynamically allocated context
+#define LOG_CONTEXT_CREATE(dest_log_context, parent_context, ...) \
+    ...
+
+// destroy a dynamically allocated context
+#define LOG_CONTEXT_DESTROY(log_context_handle) \
+    ...
+
+// macro that can be used to create a context on the stack
+#define LOG_CONTEXT_LOCAL_DEFINE(log_context, parent_context, ...) \
+    ...
+
+// when a user wants to specify properties for a context, they can use LOG_CONTEXT_PROPERTY
+// LOG_CONTEXT_PROPERTY(property_type, property_name, property_value)
```

Comment:
>LOG_CONTEXT_PROPERTY(property_type, property_name, property_value) [](http://example.com/codeflow?start=3&length=66)

One would assume that these properties are evaluated at their specification time, right? And sort-of copied... someplace until a log line is produced with these stamped properties of the context.

Can properties be replaced while SW is running?

Can the type of a property be replaced or is it defined only one?

If a property can be added to a context, how does one remove it from the context?

... I had questions.

Reply:
Clarified that contexts are immutable once built.

Reply:
Maybe a future version will allow changing them, but let's start with this that the properties are placed in the contexts at construction time (be that dynamic or on the stack).